quark/subproject.py

```diff
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
+            json.dump(self.stamp, fileobj)
+
+    def url_from_directory(self, directory, include_commit=True):
+        stamp_file = join(directory, ".stamp.quark")
+        if not exists(stamp_file):
+            raise QuarkError("Missing stamp file: " + stamp_file)
+
+        with open(stamp_file, "r") as fileobj:
+            stamp = json.load(fileobj)
+
+        if not include_commit:
+            return stamp["url"]
+
+        if not stamp["sha1"]:
+            raise QuarkError("Missing sha1 in stamp file: " + stamp_file)
+
+        # Resolve job ID if needed.
+        url = urllib.parse.urlparse(stamp["url"])
+        fragments = Subproject._parse_fragment(url) if url.fragment else {}
+        if "ref" in fragments:
+            print_msg("resolving job id for " + self.parsed_artifact_name, self._comment())
+            stamp["job_id"] = self._resolve_job_id(
+                self.parsed_project_name,
+                fragments["ref"],
+                fragments["job"],
+            )
+
+        # Generate frozen URL.
+        new_fragments = ["sha1=" + stamp["sha1"]]
+        if stamp["job_id"]:
+            new_fragments.append("job=" + stamp["job_id"])
+
+        new_url = urllib.parse.urlparse(stamp["url"])
+        new_url = new_url._replace(fragment="&".join(new_fragments))
+
+        return urllib.parse.urlunparse(new_url)
+
+    def _gitlab_setup(self, url):
+        self.gitlab_url = "https://" + url.netloc  # Enforce HTTPS for now.
+
+        # Try private token variables, in order of importance. GITLAB_PRIVATE_TOKEN is commonly used
+        # by python-gitlab's "gitlab" CLI tool, whereas GITLAB_TOKEN is used by GitLab's "glab" CLI
+        # tool.
+        for var in ("QUARK_GITLAB_PRIVATE_TOKEN", "GITLAB_PRIVATE_TOKEN", "GITLAB_TOKEN"):
+            if var in os.environ:
+                self.gitlab_token = os.environ[var]
+                self.gitlab_token_header = "PRIVATE-TOKEN"
+                self.gitlab_token_is_pat = True  # Is this a Personal/Private Access Token? (PAT)
+                return
+
+        # If we're running in a CI job, use the CI_JOB_TOKEN.
+        if "CI_JOB_TOKEN" in os.environ:
+            self.gitlab_token = os.environ["CI_JOB_TOKEN"]
+            self.gitlab_token_header = "JOB-TOKEN"
+            self.gitlab_token_is_pat = False  # Is this a Personal/Private Access Token? (PAT)
+            return
+
+        raise QuarkError("Missing authentication token.")
+
+    def _parse_url(self, url):
+        fragments = Subproject._parse_fragment(url) if url.fragment else {}
+
+        # There are mainly used for logging messages.
+        self.parsed_job = None
+        self.parsed_ref = None
+        self.parsed_pkg = None
+
+        self.stamp["sha1"] = fragments.get("sha1")
+
+        if url.scheme == "gitlab+ci":
+            if "job" not in fragments:
+                raise QuarkError("Missing job fragment in URL: " + url)
+
+            if "ref" in fragments and "job" not in fragments:
```

Comment:
Isn't the second operand of `and` already handled in the `if` immediately above?